### PR TITLE
job definition filter implementation /issue #9

### PR DESCRIPTION
**Pull request by mrdavidlaing**
This is an implementation of what @mrdavidlaing proposed in his [comment](https://github.com/logsearch/logsearch-boshrelease/issues/9#issuecomment-41587591).

Users are able to update the `logstash_parser.filters` config with URLs of their tar/tar.gz/zip files. Files are only downloaded once and are extracted in order of appearance within the array value. This means a later file could overwrite an earlier, upstream version if using the exact same name (which I consider a feature). Once extracted, they're loaded into the logstash configuration file still in order of priority based on the name.

---

Relocated from https://github.com/logsearch/logsearch-boshrelease/pull/19 to target `develop` as per #18

### CODE REVIEW

File: packages/logstash/spec [hunk 1]
```diff
@@ -3,3 +3,4 @@ name: logstash
 dependencies: []
 files: 
   - logstash/logstash-1.3.0.jar
+  - logstash/logsearch-filters-common--21e1a3b643.tgz
```
Comment (mrdavidlaing) [inline packages/logstash/spec:6]: Package logsearch-filters-common with the bosh release

File: examples/bosh-lite.yml [hunk 2]
```diff
@@ -158,6 +158,8 @@ properties:
     host: 10.244.2.10
   logstash_parser:
     debug: true
+    filters:
+    - "https://ci-logsearch.s3.amazonaws.com/logsearch-filters-cf/logstash-filters-cf-7.tgz"
```
Comment (mrdavidlaing) [inline examples/bosh-lite.yml:162]: Also allow pulling extra external filters

File: jobs/log_parser/templates/bin/log_parser_ctl [hunk 3]
```diff
@@ -20,17 +20,60 @@ case $1 in
...
+    pushd ${JOB_DIR}/config > /dev/null
+
+    <% p('logstash_parser.filters').each_with_index do | v, i | %>
+      # <%= v %>
+      if ! [[ -d raw/<%= md5filters[i] %> ]] ; then
+        mkdir -p raw/<%= md5filters[i] %>
+        pushd raw/<%= md5filters[i] %> > /dev/null
+
+        <% if v =~ /\.tar\.gz$/ or v =~ /\.tgz$/ %>
+          /usr/bin/wget -qO- "<%= v %>" | /bin/tar -xz
```
Comment (mrdavidlaing) [inline jobs/log_parser/templates/bin/log_parser_ctl:52]: I dropped the `--strip-components 1` to support filters that don't come packaged inside a subfolder

File: jobs/log_parser/templates/bin/log_parser_ctl [hunk 4]
```diff
@@ -20,17 +20,60 @@ case $1 in
...
+        <% elsif v =~ /\.tar$/ %>
+          /usr/bin/wget -qO- "<%= v %>" | /bin/tar -x 
+        <% elsif v =~ /\.zip$/ %>
+          /usr/bin/wget "<%= v %>" && /usr/bin/unzip -d .ziptmp `basename "<%= v %>"` && /bin/mv .ziptmp/*/* ./ && /bin/rm -fr `basename "<%= v %>"` .ziptmp
+        <% end %>
+
+        popd > /dev/null
+      fi
+
+      cp -R raw/<%= md5filters[i] %>/* composite/
```
Comment (mrdavidlaing) [inline jobs/log_parser/templates/bin/log_parser_ctl]: The `-R` should cover filters that are nested within subfolders

**Comment by dpb587** [comment]
Okay. I think there may be problems if multiple filtersets (package or job) have a file by the same name (they may be applied inconsistently due to file system listings and sort only using the base file name).

**Pull request merged**